File: apps/android/src/main/java/com/tilawah/android/features/profile/EditProfileViewModel.kt
package com.tilawah.android.features.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tilawah.android.app.AppError
import com.tilawah.android.backend.AuthUser
import com.tilawah.android.backend.ProfileService
import com.tilawah.android.storage.AuthSession
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Backs the Edit Profile sheet. Owns `displayName` / `level` inputs,
 * the submit lifecycle, and propagates the BFF's refreshed user shape
 * back into [AuthSession] so the Profile header repaints without a
 * re-sign-in.
 *
 * Mirrors `apps/ios/Sources/QuranRecitationApp/Features/Profile/EditProfileViewModel.swift`.
 */
class EditProfileViewModel(
    initialUser: AuthUser,
    private val profileService: ProfileService,
    private val authSession: AuthSession,
) : ViewModel() {

    private val initialDisplayName: String = (initialUser.displayName ?: "").trim()
    private val initialLevel: String? = initialUser.level

    private val _state = MutableStateFlow(
        EditProfileUiState(
            displayName = initialDisplayName,
            level = initialLevel,
            status = EditProfileStatus.Idle,
        ),
    )
    val state: StateFlow<EditProfileUiState> = _state.asStateFlow()

    fun setDisplayName(value: String) {
        _state.update { current ->
            current.copy(
                displayName = value,
                status = if (current.status is EditProfileStatus.Error) EditProfileStatus.Idle else current.status,
            )
        }
    }

    fun setLevel(value: String?) {
        _state.update { current ->
            current.copy(
                level = value,
                status = if (current.status is EditProfileStatus.Error) EditProfileStatus.Idle else current.status,
            )
        }
    }

    fun clearError() {
        _state.update { current ->
            if (current.status is EditProfileStatus.Error) {
                current.copy(status = EditProfileStatus.Idle)
            } else {
                current
            }
        }
    }

    /**
     * Returns true on success so the caller can dismiss the sheet.
     * Leaves the sheet open with an error banner on failure.
     */
    fun submit(onSuccess: () -> Unit) {
        val snapshot = _state.value
        if (!snapshot.canSubmit(initialDisplayName, initialLevel)) return

        _state.update { it.copy(status = EditProfileStatus.Submitting) }
        viewModelScope.launch {
            val trimmed = snapshot.displayName.trim()
            // `null` here means "leave the column alone" — we never echo
            // an unchanged value through the PATCH, matching iOS.
            val displayDelta = if (trimmed != initialDisplayName) trimmed else null
            val levelDelta = if (snapshot.level != initialLevel) snapshot.level else null

            try {
                val refreshed = profileService.updateProfile(
                    displayName = displayDelta,
                    level = levelDelta,
                )
                authSession.updateUser(refreshed)
                _state.update { it.copy(status = EditProfileStatus.Idle) }
                onSuccess()
            } catch (cause: AppError) {
                _state.update { it.copy(status = EditProfileStatus.Error(messageFor(cause))) }
            } catch (cause: Throwable) {
                _state.update { it.copy(status = EditProfileStatus.Error(FALLBACK_MESSAGE)) }
            }
        }
    }

    private fun messageFor(error: AppError): String = when (error) {
        is AppError.ValidationFailed -> "Please check the form and try again."
        is AppError.Network -> "Network error. Check your connection and try again."
        AppError.InvalidCredentials -> "Your session has expired. Sign in again."
        else -> FALLBACK_MESSAGE
    }

    private companion object {
        const val FALLBACK_MESSAGE = "Couldn't save changes. Try again."
    }
}

/**
 * UI state for the Edit Profile sheet. `hasChanges` and `canSubmit`
 * derive from a pair of "initial" values held by the ViewModel rather
 * than living in the state itself — the state stays a pure snapshot
 * that the Composable observes.
 */
data class EditProfileUiState(
    val displayName: String,
    val level: String?,
    val status: EditProfileStatus,
) {

    val isSubmitting: Boolean
        get() = status is EditProfileStatus.Submitting

    fun hasChanges(initialDisplayName: String, initialLevel: String?): Boolean {
        val trimmed = displayName.trim()
        return trimmed != initialDisplayName || level != initialLevel
    }

    fun canSubmit(initialDisplayName: String, initialLevel: String?): Boolean {
        if (isSubmitting) return false
        if (!hasChanges(initialDisplayName, initialLevel)) return false
        val trimmed = displayName.trim()
        // BFF zod rejects an empty displayName and we have no UX for
        // clearing a name; keep Save disabled rather than land in 400.
        if (initialDisplayName.isNotEmpty() && trimmed.isEmpty()) return false
        return true
    }
}

sealed class EditProfileStatus {
    data object Idle : EditProfileStatus()
    data object Submitting : EditProfileStatus()
    data class Error(val message: String) : EditProfileStatus()
}

/**
 * Level vocabulary accepted by the BFF zod schema
 * (`apps/bff/src/auth/users.ts:VALID_LEVELS`). Kept here as raw strings
 * so the wire mapping lives in one place.
 */
val EditProfileLevelOptions: List<String> = listOf("beginner", "intermediate", "advanced")

File: apps/ios/Sources/QuranRecitationApp/Features/Profile/EditProfileViewModel.swift
import Foundation

/// ViewModel for the Edit Profile sheet. Backs `displayName` and `level`
/// inputs, owns the submit lifecycle, and propagates the BFF's
/// refreshed user shape back into `SessionStore` so the header repaints
/// without a re-sign-in.
@MainActor
final class EditProfileViewModel: ObservableObject, Identifiable {
  /// Stable per-instance id so `.sheet(item:)` can drive the modal off
  /// "create a fresh ViewModel" rather than a bool — that pattern owns
  /// ViewModel lifetime in the parent View and avoids the @StateObject
  /// re-init pitfall.
  let id = UUID()

  /// Skill bucket options offered by the form. Mirrors the BFF zod
  /// schema (`apps/bff/src/auth/users.ts:VALID_LEVELS`). Kept here as
  /// raw strings so the wire mapping is one place and the SwiftUI
  /// Picker can iterate them directly.
  static let levelOptions: [String] = ["beginner", "intermediate", "advanced"]

  @Published var displayName: String
  /// `nil` means "no skill bucket selected" — distinct from any string
  /// value. Persisting `nil` is valid and clears the column server-
  /// side.
  @Published var level: String?
  @Published private(set) var status: Status = .idle

  enum Status: Equatable {
    case idle
    case submitting
    case error(String)
  }

  /// The persisted initial for the sheet's avatar, captured at open
  /// time so it doesn't flicker as the user edits the name field.
  let avatarInitial: String

  private let initialDisplayName: String
  private let initialLevel: String?
  private let profileService: ProfileService
  private let session: SessionStore

  init(user: AuthenticatedUser, profileService: ProfileService, session: SessionStore) {
    let trimmed = (user.displayName ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    self.displayName = trimmed
    self.level = user.level
    self.avatarInitial = ProfileHeader.initial(displayName: user.displayName, email: user.email)
    self.initialDisplayName = trimmed
    self.initialLevel = user.level
    self.profileService = profileService
    self.session = session
  }

  /// True when the form has at least one delta vs. the initial user
  /// snapshot. Drives the "Save" button's enabled state — the form
  /// won't fire a no-op PATCH against the BFF.
  var hasChanges: Bool {
    let trimmed = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed != initialDisplayName || level != initialLevel
  }

  var isSubmitting: Bool {
    if case .submitting = status { return true }
    return false
  }

  /// `Save` is enabled when the form has a real delta vs. the initial
  /// snapshot AND clearing a non-empty displayName isn't being
  /// attempted (the BFF zod schema rejects an empty string; the
  /// "remove your name" flow has no UX yet and would lock the form on
  /// a 400). Initial-nil display names stay nullable until the user
  /// types something.
  var canSubmit: Bool {
    if isSubmitting { return false }
    guard hasChanges else { return false }
    let trimmed = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    if !initialDisplayName.isEmpty && trimmed.isEmpty {
      return false
    }
    return true
  }

  /// Submit the diff to the BFF. Returns `true` on success so the
  /// caller can dismiss the sheet. `false` leaves the sheet open with
  /// the error banner.
  func submit() async -> Bool {
    status = .submitting
    let trimmed = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    // `displayName` and `level` are PATCHed only when changed —
    // ProfileService treats nil here as "leave the column alone" so a
    // user editing just their level doesn't bounce displayName through
    // the form.
    let displayDelta = trimmed != initialDisplayName ? trimmed : nil
    let levelDelta = level != initialLevel ? level : nil
    do {
      let user = try await profileService.updateProfile(
        displayName: displayDelta,
        level: levelDelta
      )
      session.updateUser(user)
      status = .idle
      return true
    } catch let error as AppError {
      status = .error(Self.message(for: error))
      return false
    } catch {
      status = .error(Self.fallbackMessage)
      return false
    }
  }

  /// Public so the sheet can reset its banner when the user edits a
  /// field. Mirrors the Web pattern of clearing the error on input.
  func clearError() {
    if case .error = status {
      status = .idle
    }
  }

  // MARK: - Error copy

  static func message(for error: AppError) -> String {
    switch error {
    case .validationFailed:
      return "Please check the form and try again."
    case .network:
      return "Network error. Check your connection and try again."
    case .invalidCredentials:
      // Shouldn't reach here on the profile-update path (no current-
      // password verification), but if AuthHTTPClient evicts the
      // session mid-flight the sheet still shows a sensible string.
      return "Your session has expired. Sign in again."
    default:
      return fallbackMessage
    }
  }

  static let fallbackMessage = "Couldn't save changes. Try again."
}
